types/src/zkid.rs

```diff
@@ -0,0 +1,429 @@
+        );
+
+        Ok(())
+    }
+
+    pub fn verify_jwt_signature(&self, rsa_jwk: RSA_JWK, jwt_header: &String) -> Result<()> {
+        let jwt_payload = &self.jwt_payload;
+        let jwt_sig = &self.jwt_sig;
+        let jwt_token = format!("{}.{}.{}", jwt_header, jwt_payload, jwt_sig);
+        rsa_jwk.verify_signature(&jwt_token)?;
+        Ok(())
+    }
+
+    pub fn reconstruct_oauth_nonce(
+        &self,
+        exp_timestamp_secs: u64,
+        epk: &EphemeralPublicKey,
+    ) -> Result<String> {
+        let mut frs = poseidon_bn254::pad_and_pack_bytes_to_scalars_with_len(
+            epk.to_bytes().as_slice(),
+            MAX_EPK_BYTES,
+        )?;
+
+        frs.push(ark_bn254::Fr::from(exp_timestamp_secs));
+        frs.push(poseidon_bn254::pack_bytes_to_one_scalar(
+            &self.epk_blinder[..],
+        )?);
+
+        let nonce_fr = poseidon_bn254::hash_scalars(frs)?;
+        let mut nonce_bytes = [0u8; NONCE_NUM_BYTES];
+        nonce_fr.serialize_uncompressed(&mut nonce_bytes[..])?;
+
+        Ok(base64::encode_config(nonce_bytes, URL_SAFE_NO_PAD))
+    }
+}
+
+impl TryFrom<&[u8]> for OpenIdSig {
+    type Error = CryptoMaterialError;
+
+    fn try_from(bytes: &[u8]) -> Result<Self, CryptoMaterialError> {
+        bcs::from_bytes::<OpenIdSig>(bytes).map_err(|_e| CryptoMaterialError::DeserializationError)
+    }
+}
+
+#[skip_serializing_none]
+#[derive(Debug, Serialize, Deserialize)]
+pub struct OidcClaims {
+    iss: String,
+    aud: String,
+    sub: String,
+    nonce: String,
+    iat: u64,
+    email: Option<String>,
+    email_verified: Option<Value>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Claims {
+    #[serde(flatten)]
+    oidc_claims: OidcClaims,
+    #[serde(default)]
+    additional_claims: BTreeMap<String, Value>,
+}
+
+impl Claims {
+    fn get_uid_val(&self, uid_key: &String) -> Result<String> {
+        match uid_key {
+            s if s == "email" => {
+                let email_verified = self
+                    .oidc_claims
+                    .email_verified
+                    .clone()
+                    .context("'email_verified' claim is missing")?;
+                // the 'email_verified' claim may be a boolean or a boolean-as-a-string.
+                let email_verified_as_bool = email_verified.as_bool().unwrap_or(false);
+                let email_verified_as_str = email_verified.as_str().unwrap_or("false");
+                ensure!(
+                    email_verified_as_bool || email_verified_as_str.eq("true"),
+                    "'email_verified' claim was not \"true\""
+                );
+                self.oidc_claims
+                    .email
+                    .clone()
+                    .context("email claim missing on jwt")
+            },
+            s if s == "sub" => Ok(self.oidc_claims.sub.clone()),
+            _ => {
+                let uid_val = self
```

Comment:
@heliuchuan, initially, we will only accept `email` or `sub`, just like our circuit (cc @mstraka100). Later on, we might decide to allow other fields.

Can you update this to reflect this?

Reply:
@mstraka100 can you confirm?

Reply:
done